server/channels/app/user.go

```diff
@@ -401,19 +401,40 @@ func (a *App) CreateOAuthUser(c request.CTX, service string, userData io.Reader,
 		return nil, err
 	}
 
-	if teamID != "" {
-		err = a.AddUserToTeamByTeamId(c, teamID, user)
+	if err = a.AddUserToTeamByInviteIfNeeded(c, ruser, inviteToken, inviteId); err != nil {
+		c.Logger().Warn("Failed to add user to team", mlog.Err(err))
+	}
+
+	return ruser, nil
+}
+
+func (a *App) AddUserToTeamByInviteIfNeeded(c request.CTX, user *model.User, inviteToken string, inviteId string) *model.AppError {
+	var team *model.Team
+	var err *model.AppError
+
+	if inviteToken != "" {
+		team, _, err = a.AddUserToTeamByToken(c, user.Id, inviteToken)
 		if err != nil {
-			return nil, err
+			c.Logger().Warn("Failed to add user to team using invite token", mlog.Err(err))
+			return err
 		}
-
-		err = a.AddDirectChannels(c, teamID, user)
+	} else if inviteId != "" {
+		team, _, err = a.AddUserToTeamByInviteId(c, inviteId, user.Id)
```

Comment:
The [AddUserToTeamByInviteId](https://github.com/mattermost/mattermost/blob/b3d0fb6b1becf7f44db9f25296d48c23058b84ad/server/channels/app/team.go#L696) function is missing the `team.IsGroupConstrained()` check that is present in the [AddUserToTeamByToken](https://github.com/mattermost/mattermost/blob/b3d0fb6b1becf7f44db9f25296d48c23058b84ad/server/channels/app/team.go#L647) function but also on the [AddTeamMemberByInviteId](https://github.com/mattermost/mattermost/blob/b3d0fb6b1becf7f44db9f25296d48c23058b84ad/server/channels/app/team.go#L1092-L1094) which is the only place where `AddUserToTeamByInviteId` was called up till now.